flink-runtime/src/main/java/org/apache/flink/runtime/operators/JoinDriver.java

```diff
@@ -128,17 +130,22 @@ public void prepare() throws Exception{
 				ConfigConstants.DEFAULT_RUNTIME_HASH_JOIN_BLOOM_FILTERS);
 
 		// create and return joining iterator according to provided local strategy.
-		if (objectReuseEnabled) {
-			switch (ls) {
-				case INNER_MERGE:
-					this.joinIterator = new ReusingMergeInnerJoinIterator<>(in1, in2, 
+		if (reset) {
+			resetForIterativeTasks(in1, in2, serializer1, serializer2, comparator1, comparator2, pairComparatorFactory);
+			reset = false;
+		}
+		if (joinIterator == null) {
```

Comment:
This solution here with the `reset` flag is very hacky. I would like to point your attention to the `initialize` method of `ResettableDriver`, which is called once before the first iteration step. Instead of controlling with flags what `prepare` does, you should cleanly separate its work into `initialize` and `reset`. (If they have overlapping parts, then these can go to a new private method that both of them call.)